zorg/jenkins/jobs/util/run_lldb_metrics.sh

```diff
@@ -0,0 +1,180 @@
+    local stats_filename="${CSV_DIR}/${test_case_name}.json"
+
+    # Launch hyperfine.
+    profile \
+        $test_case_name \
+        "${profile_invocation} -o 'script -- f=open(\"${stats_filename}\",\"w\"); lldb.debugger.SetOutputFileHandle(f,True); lldb.debugger.HandleCommand(\"statistics dump\")' --batch -- ${debuggee_invocation}"
+
+    [[ -f $stats_filename ]] && cat $stats_filename
+}
+
+# Clean previous results
+
+rm -rf $ARTIFACTS
+mkdir $ARTIFACTS
+mkdir $CSV_DIR
+
+cat >${TEST_SRC_FILE} <<EOL
+#include <map>
+
+int fib(int n) {
+    static auto cache = [] {
+        auto ans = std::map<int, int>();
+        ans[0] = 0;
+        ans[1] = 1;
+        return ans;
+    }();
+    if (auto it = cache.find(n); it != cache.end()) return it->second;
+    auto ans = fib(n - 1) + fib(n - 2);
+    cache[ans] = ans;
+    return ans;
+}
+
+int main() {
+    fib(5);
+    return 0;
+}
+EOL
+
+# Clang benchmarks
+
+profile_clang \
+    "frame_status" \
+    "${LLVM_BUILD_DIR}/bin/lldb -o 'b CodeGenFunction::GenerateCode' -o run"
+
+profile_clang \
+    "expr_deref" \
+    "${LLVM_BUILD_DIR}/bin/lldb -o 'b CodeGenFunction::GenerateCode' -o run -o 'expr *Fn'"
+
+profile_clang \
+    "expr_method_call" \
+    "${LLVM_BUILD_DIR}/bin/lldb -o 'b CodeGenFunction::GenerateCode' -o run -o 'expr Fn->isVarArg()'"
+
+profile_clang \
+    "expr_re_eval" \
+    "${LLVM_BUILD_DIR}/bin/lldb -o 'b CodeGenFunction::GenerateCode' -o run -o 'expr *Fn' -o 'expr *Fn'"
+
+profile_clang \
+    "expr_two_stops" \
+    "${LLVM_BUILD_DIR}/bin/lldb -o 'b CodeGenFunction::GenerateCode' -o run -o 'expr *Fn' -o 'br del 1' -o 'b AsmPrinter::emitFunctionBody' -o c -o 'expr this->isVerbose()' -o 'expr TM'"
+
+profile_clang \
+    "var_then_expr" \
+    "${LLVM_BUILD_DIR}/bin/lldb -o 'b CodeGenFunction::GenerateCode' -o run -o up -o 'frame var' -o down -o 'expr *this'"
+
+profile_clang \
+    "var" \
+    "${LLVM_BUILD_DIR}/bin/lldb -o 'b CodeGenFunction::GenerateCode' -o run -o 'frame var'"
+
+# LLDB benchmarks
+
+profile_lldb \
+    "frame_status" \
+    "${LLVM_BUILD_DIR}/bin/lldb -o 'b ClangASTSource::FindExternalVisibleDeclsByName' -o run"
+
+profile_lldb \
+    "expr_deref" \
+    "${LLVM_BUILD_DIR}/bin/lldb -o 'b ClangASTSource::FindExternalVisibleDeclsByName' -o run -o 'expr m_active_lexical_decls'"
+
+profile_lldb \
+    "expr_method_call" \
+    "${LLVM_BUILD_DIR}/bin/lldb -o 'b ClangASTSource::FindExternalVisibleDeclsByName' -o run -o 'expr clang_decl_name.getAsString()'"
+
+profile_lldb \
+    "expr_re_eval" \
+    "${LLVM_BUILD_DIR}/bin/lldb -o 'b ClangASTSource::FindExternalVisibleDeclsByName' -o run -o 'expr m_active_lexical_decls' -o 'expr m_active_lexical_decls'"
+
+profile_lldb \
+    "expr_two_stops" \
+    "${LLVM_BUILD_DIR}/bin/lldb -o 'b ClangASTSource::FindExternalVisibleDeclsByName' -o run -o 'expr m_active_lexical_decls' -o 'b CodeGeneratorImpl::HandleTranslationUnit' -o 'br del 1' -o c -o 'expr Builder' -o 'expr Ctx.getLangOpts()'"
+
```

Comment:
Could you add some comments explaining for future maintainers why these various breakpoints and expressions were chosen and what is being tested?

For example, I expect the first one to trigger all sorts of symbol file loading, the second one to test the same expression with a warm cache, one finding a templated type with lots of false positives that appears 1000 times, ...

Reply:
Took a stab at it. Wasn't sure of the level of detail. Let me know if i should elaborate more here